library/src/com/google/maps/android/BiMultiMap.java

```diff
@@ -0,0 +1,58 @@
+package com.google.maps.android;
+
+import java.util.Collection;
+import java.util.HashMap;
+
+/**
+ * Extension of HashMap that provides two main features. Firstly it allows reverse lookup
+ * for a key given a value, by storing a second HashMap internally which maps values to keys.
+ * Secondly, it supports Collection values, in which case, each item in the collection is
+ * used as a key in the internal reverse HashMap. It's therefore up to the caller to ensure
+ * the overall set of values, and collection values, are unique.
+ *
+ * Used by GeoJsonRenderer to store GeoJsonFeature instances mapped to corresponding Marker,
+ * Polyline, and Polygon map objects. We want to look these up in reverse to provide access
+ * to GeoJsonFeature instances when map objects are clicked.
+ */
+public class BiMultiMap<K> extends HashMap<K, Object> {
+
+    private final HashMap<Object, K> mValuesToKeys = new HashMap<>();
+
+    public void putAll(HashMap<K, Object> map) {
+        // put() manages the reverse map, so call it on each entry.
+        for (Entry<K, Object> entry : map.entrySet()) {
+            put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    @Override
+    public Object put(K key, Object value) {
+        // Store value/key in the reverse map, and store each item in the value
+        // if the value is a collection.
+        if (value instanceof Collection) {
```

Comment:
You could probably do this a little cleaner with a separate method. e.g.

    public Object put(K key, Collection value) ...